examples/api/lib/material/icon_button/icon_button.2.dart

```diff
@@ -0,0 +1,117 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flutter code sample for IconButton
+
+import 'package:flutter/material.dart';
+
+void main() {
+  runApp(const IconButtonApp());
+}
+
+class IconButtonApp extends StatelessWidget {
+  const IconButtonApp({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      theme: ThemeData(colorSchemeSeed: const Color(0xff6750a4), useMaterial3: true),
+      title: 'Icon Button Types',
+      home: const Scaffold(
+        body: ButtonTypesExample(),
+      ),
+    );
+  }
+}
+
+class ButtonTypesExample extends StatelessWidget {
+  const ButtonTypesExample({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return Padding(
+      padding: const EdgeInsets.all(4.0),
+      child: Row(
+        children: const <Widget>[
+          Spacer(),
+          ButtonTypesGroup(enabled: true),
+          ButtonTypesGroup(enabled: false),
+          Spacer(),
+        ],
+      ),
+    );
+  }
+}
+
+class ButtonTypesGroup extends StatelessWidget {
+  const ButtonTypesGroup({ super.key, required this.enabled });
+
+  final bool enabled;
+
+  @override
+  Widget build(BuildContext context) {
+    final VoidCallback? onPressed = enabled ? () {} : null;
+    final ColorScheme colors = Theme.of(context).colorScheme;
+
+    return Padding(
+      padding: const EdgeInsets.all(4.0),
+      child: Column(
+        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
+        children: <Widget>[
+          IconButton(icon: const Icon(Icons.filter_drama), onPressed: onPressed),
+
+          // Use a standard IconButton with specific style to implement the
+          // 'Filled' type.
+          IconButton(
+            icon: const Icon(Icons.filter_drama),
+            onPressed: onPressed,
+            style: IconButton.styleFrom(
+              foregroundColor: colors.onPrimary,
+              backgroundColor: colors.primary,
+              disabledBackgroundColor: colors.onSurface.withOpacity(0.12),
+              hoverColor: colors.onPrimary.withOpacity(0.08),
+              focusColor: colors.onPrimary.withOpacity(0.12),
+              highlightColor: colors.onPrimary.withOpacity(0.12),
```

Comment:
We should use tokens for these. We should also provide the three default styles, so that devs don't have to copy paste and have a nice starting point.

Reply:
Yeah, we have a similar problem with the variants in the common buttons as well. I didn't realize the variants would need that much configuration. 

I really wish we had a simpler way of providing these variants without a lot of extra API baggage to go with them (i.e. either full button subclasses for each variant, or named constructors that have to duplicate all the constructor parameters and use an enum or something to indicate which set of defaults to use in the build method).

That said, I would go with this example for this PR and we can look at cleaning up the variant support in a future PR.

Reply:
With the current Flutter master channel (91603595d1), `IconButton.styleFrom` gives me stadium shapes by default.  This is consistent with [icon_button.dart#L614](https://github.com/flutter/flutter/pull/105176/files#diff-e6dbc7e9fe24c98802742647210164f9ea54a6ff8ba15b87eb9ee5d1ce758482R614), below.

I believe the spec for [`md.sys.shape.corner.full`](https://m3.material.io/m3/pages/shape/shape-scale-tokens#6f668ba1-b671-4ea2-bcf3-c1cff4f4099e) indicates they should use `shape: const CircleBorder()` instead.

Reply:
Hi, the icon button is migrated to be a subclass of ButtonStyleButton now and has a compact `VisualDensity` on desktop platform, so the buttons don't look like circular. If the app is ran on iOS/Android, the buttons will show a standard `VisualDensity` and have a circular shape.

If this is the problem that you are facing, adding `visualDensity: VisualDensity.standard` to IconButton or IconButton.styleFrom() will solve the problem on desktop platform:)

Reply:
@QuncCccccc Interesting, thanks!  I didn't appreciate that property.  The spec does have a ["high-emphasis filled button for End call"](https://m3.material.io/components/icon-buttons/guidelines) in an example, but it's a bit subtle.

For reference, I see the issue in Flutter web.

Reply:
Sorry, I assumed you are using a desktop platform. Both desktop and web have a compact visual density as a default value because they have a larger screen, compared to phone devices. So on web, just adding same thing as desktop (`visualDensity: VisualDensity.standard`) will provide a circular icon button:)

Please let me know if there's any other problems.